File: packages/py_agents/my_agent_project_py_agents/py_http_agent/agent.py
from contextlib import contextmanager

from my_agent_project_agent_connection import PyA2aAgentClientStrands, PyAgentsMcpServerClientStrands, log_model_errors
from strands import Agent, tool
from strands_tools import current_time


@tool
def subtract(a: int, b: int) -> int:
    return a - b


@contextmanager
def get_agent():
    py_agents_mcp_server = PyAgentsMcpServerClientStrands.create()
    with (
        py_agents_mcp_server,
    ):
        py_a2a_agent = PyA2aAgentClientStrands.create()

        @tool
        def ask_py_a2a_agent(prompt: str) -> str:
            """Delegate a question to the remote PyA2aAgent A2A agent and return its reply."""
            return str(py_a2a_agent(prompt))

        yield Agent(
            name="PyHttpAgent",
            description="PyHttpAgent Strands Agent",
            system_prompt="""
You are a mathematical wizard.
Use your tools for mathematical tasks.
Refer to tools as your 'spellbook'.
""",
            tools=[subtract, current_time, ask_py_a2a_agent, *py_agents_mcp_server.list_tools_sync()],
            hooks=[log_model_errors],
        )

File: packages/agents/src/http-agent/agent.ts
import { Agent, tool } from '@strands-agents/sdk';
import {
  logModelErrors,
  logToolErrors,
  A2aAgentClientStrands,
  AgentsMcpServerClientStrands,
} from '@my-agent-project/agent-connection';
import { getSessionManager } from './session.js';
import { z } from 'zod';

const multiply = tool({
  name: 'Multiply',
  description: 'Multiply two numbers',
  inputSchema: z.object({
    a: z.number(),
    b: z.number(),
  }),
  callback: ({ a, b }) => a * b,
});

export const getAgent = async () => {
  const agentsMcpServer = await AgentsMcpServerClientStrands.create();
  const a2aAgent = await A2aAgentClientStrands.create();
  const a2aAgentTool = tool({
    name: 'askA2aAgent',
    description:
      'Delegate a question to the remote A2aAgent A2A agent and return its reply.',
    inputSchema: z.object({ prompt: z.string() }),
    callback: async ({ prompt }) => (await a2aAgent.invoke(prompt)).toString(),
  });
  const agent = new Agent({
    sessionManager: await getSessionManager(),
    systemPrompt: `You are a mathematical wizard.
  Use your tools for mathematical tasks.
  Refer to tools as your 'spellbook'.`,
    tools: [agentsMcpServer, multiply, a2aAgentTool],
  });
  logModelErrors(agent);
  logToolErrors(agent);
  return agent;
};
